presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java

```diff
@@ -0,0 +1,419 @@
+ * https://docs.google.com/document/d/18HN7peS2eR8lZsErqcmnoWyMEPb6p4OQeidH1JP_EkA/edit#heading=h.hb6ihepq7q74
+ * <p>
+ * From:
+ * - Apply (with correlation list: [C])
+ * - (input) plan which produces symbols: [A, B, C]
+ * - (subquery) Aggregation(GROUP BY 1; functions: [sum(F),...]
+ * - Filter(D = C AND E > 5)
+ * - plan which produces symbols: [D, E, F]
+ * to:
+ * - Aggregation(GROUP BY A, B, C, U; functions: [sum(F),...]
+ * - Join(LEFT_OUTER, D = C)
+ * - AssignUniqueId(adds symbol U)
+ * - (input) plan which produces symbols: [A, B, C]
+ * - Filter(E > 5)
+ * - plan which produces symbols: [D, E, F]
+ * <p>
+ * Note only conjunction predicates in FilterNode are supported
+ */
+public class TransformCorrelatedScalarAggregationToJoin
+        implements PlanOptimizer
+{
+    private final Metadata metadata;
+
+    public TransformCorrelatedScalarAggregationToJoin(Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+    }
+
+    @Override
+    public PlanNode optimize(
+            PlanNode plan,
+            Session session,
+            Map<Symbol, Type> types,
+            SymbolAllocator symbolAllocator,
+            PlanNodeIdAllocator idAllocator)
+    {
+        return rewriteWith(new Rewriter(idAllocator, symbolAllocator, metadata), plan, null);
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<PlanNode>
+    {
+        private final PlanNodeIdAllocator idAllocator;
+        private final SymbolAllocator symbolAllocator;
+        private final Metadata metadata;
+
+        public Rewriter(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Metadata metadata)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.symbolAllocator = requireNonNull(symbolAllocator, "symbolAllocator is null");
+            this.metadata = requireNonNull(metadata, "metadata is null");
+        }
+
+        @Override
+        public PlanNode visitApply(ApplyNode node, RewriteContext<PlanNode> context)
+        {
+            ApplyNode rewrittenNode = (ApplyNode) context.defaultRewrite(node, context.get());
+            if (!rewrittenNode.getCorrelation().isEmpty()) {
+                Optional<AggregationNode> scalarAggregation = searchFrom(rewrittenNode.getSubquery())
+                        .where(AggregationNode.class::isInstance)
+                        .skipOnlyWhen(isInstanceOfAny(ProjectNode.class, EnforceSingleRowNode.class))
+                        .findFirst();
+                if (scalarAggregation.isPresent() && scalarAggregation.get().getGroupBy().isEmpty()) {
+                    return rewriteScalarAggregation(rewrittenNode, scalarAggregation.get());
+                }
+            }
+            return rewrittenNode;
+        }
+
+        private PlanNode rewriteScalarAggregation(ApplyNode applyNode, AggregationNode scalarAggregation)
```

Comment:
Calling the arguments`apply` and `aggregation`, respectively, may improve readability slightly.